osu.Game/Skinning/SkinProvidingContainer.cs

```diff
@@ -53,177 +41,193 @@ public class SkinProvidingContainer : Container, ISkinSource
+        }
 
-            SkinSources.BindCollectionChanged(((_, args) =>
-            {
-                switch (args.Action)
-                {
-                    case NotifyCollectionChangedAction.Add:
-                        foreach (var skin in args.NewItems.Cast<ISkin>())
-                        {
-                            disableableSkinSources.Add(skin, new DisableableSkinSource(skin, this));
-
-                            if (skin is ISkinSource source)
-                                source.SourceChanged += OnSourceChanged;
-                        }
-
-                        break;
-
-                    case NotifyCollectionChangedAction.Reset:
-                    case NotifyCollectionChangedAction.Remove:
-                        foreach (var skin in args.OldItems.Cast<ISkin>())
-                        {
-                            disableableSkinSources.Remove(skin);
-
-                            if (skin is ISkinSource source)
-                                source.SourceChanged -= OnSourceChanged;
-                        }
-
-                        break;
-
-                    case NotifyCollectionChangedAction.Replace:
-                        foreach (var skin in args.OldItems.Cast<ISkin>())
-                        {
-                            disableableSkinSources.Remove(skin);
+        /// <summary>
+        /// Add a new skin to this provider. Will be added to the end of the lookup order precedence.
+        /// </summary>
+        /// <param name="skin">The skin to add.</param>
+        protected void AddSource(ISkin skin)
+        {
+            skinSources.Add(skin, new DisableableSkinSource(skin, this));
 
-                            if (skin is ISkinSource source)
-                                source.SourceChanged -= OnSourceChanged;
-                        }
+            if (skin is ISkinSource source)
+                source.SourceChanged += TriggerSourceChanged;
+        }
 
-                        foreach (var skin in args.NewItems.Cast<ISkin>())
-                        {
-                            disableableSkinSources.Add(skin, new DisableableSkinSource(skin, this));
+        /// <summary>
+        /// Remove a skin from this provider.
+        /// </summary>
+        /// <param name="skin">The skin to remove.</param>
+        protected void RemoveSource(ISkin skin)
+        {
+            if (!skinSources.Remove(skin))
+                return;
 
-                            if (skin is ISkinSource source)
-                                source.SourceChanged += OnSourceChanged;
-                        }
+            if (skin is ISkinSource source)
+                source.SourceChanged -= TriggerSourceChanged;
+        }
 
-                        break;
-                }
-            }), true);
+        /// <summary>
+        /// Clears all skin sources.
+        /// </summary>
+        protected void ResetSources()
+        {
+            foreach (var skin in AllSources.ToArray())
```

Comment:
```suggestion
            foreach (var skin in skinSources.Key.ToArray())
```

Reply:
Can you explain what you see here? ie. why you are proposing this change?

Reply:
just noticed it while reading, the previous one doesn't seem to cause any behaviour change but either way the suggestion is correcting it.

Reply:
I guess because it was the only local usage of the public property? I've updated it.